### ISSUE TITLE
Make dune works with metaocaml

### ISSUE THREAD

**Issue by Drup** [ocaml/dune#1413]
I was doing some experiments with metaocaml, and managing to make it work with dune was ... annoying.

The way metaocaml is currently setup is that, to use metaocaml code, you need to use the ocamlfind toolchain "metaocaml" that is installed by `base-metaocaml-ocamlfind` opam package. This is true even if you use a separate switch for metaocaml.

I tried to use the toolchain feature of dune, with `-x`. That started out promising but didn't work out so well, because dune still insisted on using the default toolchain to build executables, which prevented linking `runcode`. 

In the end, I ended up doing a rather violent `ln -sf metaocamlc ocamlc.opt`. I couldn't make the native version work properly.

cc @yallop, as I expect him to know more about this whole toolchain setup (which I would really like to see gone, but there are some technical difficulties ...).

**Comment by ejgallego** [comment]
Oh indeed a bit more of information would be lovely. We are using metaocaml by adding this to the link flags:
```
 (mode byte)
 (link_flags compiler-libs/ocamlcommon.cma metalib.cma)
```
It is a terrible hack I know but it works on `4.04.0+BER`.

**Comment by Drup** [comment]
Have you managed to make `Runnative` works ?

**Comment by ejgallego** [comment]
> Have you managed to make `Runnative` works ?

Likely not, as it doesn't appear in the codebase.

**Comment by yallop** [comment]
I'm not using Dune with MetaOCaml at the moment, so don't have any direct advice.  It might be possible to use something similar to the `link_flags` arrangement for `metaocamlopt`, too, though.  In N107 the implementations of `metaocamlc` and `metaocamlopt` are merged, [only distinguished by an `#ifdef`](https://github.com/metaocaml/ber-metaocaml/blob/7fccae53ba55fb533605979d5f1c45d816d159e9/ber-metaocaml-107/metaocamlc.c#L44).

**Comment by rgrinberg** [comment]
It would be really cool if dune had full support for metaocaml, as an initial solution however, dune should simply support toolchain selections.

@Drup did you give the `OCAMLFIND_TOOLCHAIN` env var a try? We could simply add support for setting this env var via the workspace file if it works.

**Comment by cyberhuman** [comment]
I just tried `OCAMLFIND_TOOLCHAIN` and it does not work.
Looking at the dune's source code, it does not use findlib, but instead it just looks inside the `$PATH` for `ocamlc.opt`.

**Comment by cyberhuman** [comment]
I figured ocamlfind has a bug in `ocamlfind printconf conf` that leads to dune using the default findlib config even when a different toolchain was selected. I have submitted a fix: https://gitlab.camlcity.org/gerd/lib-findlib/merge_requests/19

**Comment by rgrinberg** [comment]
Is that enough though? The toolchain can be defined via:

```
ocamlc(tlc) = ...
...
```

What would you expect `$ ocamlfind print conf -toolchain tlc` to print for example?

**Comment by cyberhuman** [comment]
`ocamlfind` looks for the findlib config inside lib/findlib.conf.d if toolchain is set. For example, for metaocaml it looks for lib/findlib.conf.d/metaocaml.conf, so that `ocamlfind ocamlc` invokes `metaocamlc`. I checked and it works for me (though I have some other issues with metaocaml itself though).

**Comment by MarisaKirisame** [comment]
what is the status on this? i am trying to use metaocaml with dune, and want to know if it will be supported soon.

**Comment by diml** [comment]
Reading this thread again, it seems to me that all we need to do is add support for selecting an alternative toolchain in dune, but not for cross compilation. i.e. we should add a `--toolchain` command line argument, as well as `(toolchain ...)` field in build context configuration, etc... Then a user will be able to use metaocaml by calling `dune build --toolchain metaocaml`. @rgrinberg does that seem right to you as well?

**Comment by rgrinberg** [comment]
I'm in favor of toolchain support because I do want us to have parity with findlib. But it's still feels weird to me that this required to support metaocaml. I don't quite understand why metaocaml simply can't be its own switch. Why does it need the host at all?

**Comment by diml** [comment]
Indeed, I'm wondering what the reason is as well.

**Comment by Drup** [comment]
@rgrinberg You might want to ask @yallop about that, but iirc, the idea is that the normal compiler, `ocamlc/opt`, doesn't bundle all the extra stuff need for runtime compilation. It's only brought in when using `metaocamlc/opt`. Honestly, I'm not sure why this isn't provided as an after-the-fact ocamlfind libraries, just like bigarray/unix or as a compiler-option-turned-library, like `thread`.

**Comment by yallop** [comment]
> Honestly, I'm not sure why this isn't provided as an after-the-fact ocamlfind libraries, just like bigarray/unix or as a compiler-option-turned-library, like `thread`.

I think things probably could be done that way.  It seems worth looking at for a future MetaOCaml release.

The toolchain approach is designed to work with MetaOCaml as it currently ships -- i.e. with a full OCaml-compatible toolchain (`ocaml`/`ocamlc`/`ocamlopt`), and some extra top-level tools (`metaocaml`/`metaocamlc`/`metaocamlopt`) that load the extra libraries needed to support quotation.

**Comment by Drup** [comment]
@yallop In my opinion, it would be very worthwhile, as it would allow to reuse all the normal tools (dune, utop ...) transparently. The extra `meta*` tools don't really have any reason to exist, they only make usage more annoying/complicated.

**Comment by yallop** [comment]
A `META` file like the following (lightly tested) seems to do what's needed:

```
description = "Extra libraries for MetaOCaml"
requires = "compiler-libs.common"
requires(byte) = "compiler-libs.bytecomp compiler-libs.toplevel"
requires(native) = "compiler-libs.optcomp dynlink"
archive(byte) = "+metalib.cma +berstart.cmo"
archive(native) = "+codelib.cmx +runnative.cmx"
```

i.e. it becomes possible to build and run MetaOCaml programs using the standard toolchain by passing `-package metalib` to `ocamlfind`.

**Comment by Drup** [comment]
Oh, that's really nice. I would have expected some changes in metaocaml as well for this to work. Could you add that to your `base-metaocaml-ocamlfind` package ?

**Comment by yallop** [comment]
@Drup: see https://github.com/ocaml/opam-repository/pull/13474, which is lightly tested with OCamlbuild/OCamlfind using `package(metaocaml), open(Codelib)`.

**Comment by Et7f3** [comment]
> would be nice if one could use bucklescript as a compiler with esy and dune though, having one main-workflow for both native and js
@lessp

I agree with him. Also some people in reason world have nearly managed to build bucklescirpt compiler in esy so we could maybe use it with dune if it support native toolchain.
